Library/Homebrew/linkage_checker.rb

```diff
@@ -0,0 +1,121 @@
+require "set"
+require "keg"
+require "formula"
+
+class LinkageChecker
+  attr_reader :keg, :formula
+  attr_reader :brewed_dylibs, :system_dylibs, :broken_dylibs, :variable_dylibs
+  attr_reader :undeclared_deps, :reverse_links
+
+  def initialize(keg, formula = nil)
+    @keg = keg
+    @formula = formula || resolve_formula(keg)
+    @brewed_dylibs = Hash.new { |h, k| h[k] = Set.new }
+    @system_dylibs = Set.new
+    @broken_dylibs = Set.new
+    @variable_dylibs = Set.new
+    @undeclared_deps = []
+    @reverse_links = Hash.new { |h, k| h[k] = Set.new }
+    check_dylibs
+  end
+
+  def check_dylibs
+    @keg.find do |file|
+      next if file.symlink? || file.directory?
+      next unless file.dylib? || file.mach_o_executable? || file.mach_o_bundle?
+      file.dynamically_linked_libraries.each do |dylib|
+        @reverse_links[dylib] << file
+        if dylib.start_with? "@"
+          @variable_dylibs << dylib
+        else
+          begin
+            owner = Keg.for Pathname.new(dylib)
+          rescue NotAKegError
+            @system_dylibs << dylib
+          rescue Errno::ENOENT
+            @broken_dylibs << dylib
+          else
+            @brewed_dylibs[owner.name] << dylib
+          end
+        end
+      end
+    end
+
+    @undeclared_deps = check_undeclared_deps if formula
+  end
+
+  def check_undeclared_deps
+      filter_out = proc do |dep|
+        dep.build? || (dep.optional? && !dep.option_names.any? { |n| formula.build.with?(n) })
```

Comment:
Can you split this logic into multiple `next` statements so you don't need to process the binary logic in your head? Thanks.